Content.Server/Chemistry/Containers/EntitySystems/SolutionContainerSystem.cs

```diff
@@ -0,0 +1,196 @@
+        {
+            yield return (name, solution);
+        }
+    }
+
+
+    public Solution EnsureSolution(Entity<MetaDataComponent?> entity, string name)
+        => EnsureSolution(entity, name, out _);
+
+    public Solution EnsureSolution(Entity<MetaDataComponent?> entity, string name, out bool existed)
+        => EnsureSolution(entity, name, FixedPoint2.Zero, out existed);
+
+    public Solution EnsureSolution(Entity<MetaDataComponent?> entity, string name, FixedPoint2 minVol, out bool existed)
+    {
+        var (uid, meta) = entity;
+        DebugTools.Assert(Resolve(uid, ref meta), $"Attempted to ensure solution on invalid entity {ToPrettyString(entity.Owner)}");
+
+        if (meta.EntityLifeStage >= EntityLifeStage.MapInitialized)
+            return EnsureSolutionEntity(uid, name, minVol, out existed).Comp.Solution;
+        else
+            return EnsureSolutionPrototype(uid, name, minVol, out existed);
+    }
+
+    public Entity<SolutionComponent> EnsureSolutionEntity(Entity<SolutionContainerComponent?> entity, string name, FixedPoint2 minVol, out bool existed)
+    {
+        existed = true;
+
+        var (uid, container) = entity;
+        if (!Resolve(uid, ref container, logMissing: false))
+        {
+            container = AddComp<SolutionContainerComponent>(uid);
+            existed = false;
+        }
+
+        if (!container.Solutions.TryGetValue(name, out var solutionSlot))
+        {
+            solutionSlot = ContainerSystem.EnsureContainer<ContainerSlot>(entity.Owner, $"solution@{name}");
+            container.Solutions.Add(name, solutionSlot);
+            Dirty(uid, container);
+        }
+
+        var needsInit = false;
+        SolutionComponent solutionComp;
+        if (solutionSlot.ContainedEntity is not { } solutionId)
+        {
+            (solutionId, solutionComp) = SpawnSolutionUninitialized(uid, name, new Solution() { MaxVolume = minVol, Name = name });
+            ContainerSystem.Insert(solutionId, solutionSlot, force: true);
+            existed = false;
+            needsInit = true;
+        }
+        else if (!TryComp(solutionId, out solutionComp!))
+        {
+            solutionComp = new SolutionComponent()
+            {
+                Solution = new Solution() { MaxVolume = minVol, Name = name },
+            };
+            AddComp(solutionId, solutionComp);
+            existed = false;
+        }
+        else
+        {
+            solutionComp.Solution.MaxVolume = FixedPoint2.Max(solutionComp.Solution.MaxVolume, minVol);
+            Dirty(solutionId, solutionComp);
+        }
+
+        if (!TryComp(solutionId, out ContainerSolutionComponent? relation))
+        {
+            relation = new ContainerSolutionComponent() { Container = uid, Name = name };
+            AddComp(solutionId, relation);
+        }
+        else
+        {
+            (relation.Container, relation.Name) = (uid, name);
+            Dirty(solutionId, relation);
+        }
```

Comment:
This already gets set by `SpawnSolutionUninitialized()`, which should be the only way that a solution entity is ever spawned. So for the sake of performance it'd be better to just remove this or make it an assert in a debug block.